File: src/main/java/main/commands/slashcommands/commands/misc/reminders/jobs/ReminderJob.java
package main.commands.slashcommands.commands.misc.reminders.jobs;

import main.constants.Toggles;
import main.main.Robertify;
import main.utils.GeneralUtils;
import main.utils.RobertifyEmbedUtils;
import main.utils.json.reminders.RemindersConfig;
import main.utils.json.toggles.TogglesConfig;
import main.utils.locale.LocaleManager;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.internal.utils.tuple.Pair;
import org.quartz.Job;
import org.quartz.JobExecutionContext;
import org.quartz.JobExecutionException;

import java.time.Instant;

public class ReminderJob implements Job {

    @Override
    public void execute(JobExecutionContext jobExecutionContext) throws JobExecutionException {
        final var dataMap = jobExecutionContext.getJobDetail().getJobDataMap();
        final var guildID = dataMap.getLong("guild");
        final var destination = dataMap.getLong("destination");
        final var user = dataMap.getLong("user");
        final var reminder = dataMap.getString("reminder");

        final Guild guild = Robertify.shardManager.getGuildById(guildID);

        if (guild == null) return;

        if (destination == -1L) {
            dmReminder(guild, user, reminder);
            return;
        }

        final var remindersConfig = new RemindersConfig(guild);
        if (remindersConfig.userIsBanned(user)) {
            dmReminder(guild, user, reminder);
            return;
        }

        final TextChannel channel = guild.getTextChannelById(destination);

        if (channel == null) {
            dmReminder(guild, user, reminder);
            return;
        }

        if (!guild.getSelfMember().hasPermission(channel, Permission.MESSAGE_SEND)) {
            dmReminder(guild, user, reminder);
            return;
        }

        if (!remindersConfig.channelIsBanned(channel.getIdLong())) {
            dmReminder(guild, user, reminder);
            return;
        }

        if (!new TogglesConfig(guild).getToggle(Toggles.REMINDERS))
            return;

        final var localeManager = LocaleManager.getLocaleManager(guild);
        channel.sendMessage(localeManager.getMessage(RobertifyLocaleMessage.ReminderMessages.REMINDER_SEND, Pair.of("{user}", GeneralUtils.toMention(guild, user, GeneralUtils.Mentioner.USER))))
                .setEmbeds(RobertifyEmbedUtils.embedMessageWithTitle(guild, localeManager.getMessage(RobertifyLocaleMessage.ReminderMessages.REMINDERS_EMBED_TITLE), reminder)
                        .setTimestamp(Instant.now())
                        .build())
                .queue();
    }

    private void dmReminder(Guild guild, long user, String reminder) {
        final var localeManager = LocaleManager.getLocaleManager(guild);

        GeneralUtils.dmUser(user, RobertifyEmbedUtils.embedMessageWithTitle(guild, localeManager.getMessage(RobertifyLocaleMessage.ReminderMessages.REMINDERS_EMBED_TITLE), reminder)
                .setFooter(localeManager.getMessage(RobertifyLocaleMessage.ReminderMessages.REMINDER_FROM, Pair.of("{server}", guild.getName())))
                .setTimestamp(Instant.now())
                .build()
        );
    }
}

File: src/main/kotlin/main/utils/json/reminders/scheduler/ReminderJob.kt
package main.utils.json.reminders.scheduler

import kotlinx.coroutines.runBlocking
import main.constants.Toggle
import main.main.Robertify
import main.utils.GeneralUtils
import main.utils.RobertifyEmbedUtils
import main.utils.json.reminders.RemindersConfig
import main.utils.json.toggles.TogglesConfig
import main.utils.locale.LocaleManager
import main.utils.locale.messages.ReminderMessages
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Guild
import org.quartz.Job
import org.quartz.JobDataMap
import org.quartz.JobExecutionContext
import java.time.Instant

class ReminderJob : Job {
    override fun execute(context: JobExecutionContext) {
        runBlocking {
            val dataMap: JobDataMap = context.jobDetail.jobDataMap
            val guildID = dataMap.getLong("guild")
            val destination = dataMap.getLong("destination")
            val user = dataMap.getLong("user")
            val reminder = dataMap.getString("reminder")

            val guild = Robertify.shardManager.getGuildById(guildID) ?: return@runBlocking

            if (destination == -1L) {
                dmReminder(guild, user, reminder)
                return@runBlocking
            }

            val remindersConfig = RemindersConfig(guild)
            if (remindersConfig.userIsBanned(user)) {
                dmReminder(guild, user, reminder)
                return@runBlocking
            }

            val channel = guild.getTextChannelById(destination)

            if (channel == null) {
                dmReminder(guild, user, reminder)
                return@runBlocking
            }

            if (!guild.selfMember.hasPermission(channel, Permission.MESSAGE_SEND)) {
                dmReminder(guild, user, reminder)
                return@runBlocking
            }

            if (remindersConfig.channelIsBanned(channel.idLong)) {
                dmReminder(guild, user, reminder)
                return@runBlocking
            }

            if (!TogglesConfig(guild).getToggle(Toggle.REMINDERS)) return@runBlocking

            val localeManager = LocaleManager[guild]
            channel.sendMessage(
                localeManager.getMessage(
                    ReminderMessages.REMINDER_SEND,
                    Pair(
                        "{user}",
                        GeneralUtils.toMention(guild, user, GeneralUtils.Mentioner.USER)
                    )
                )
            )
                .setEmbeds(
                    RobertifyEmbedUtils.embedMessageWithTitle(
                        guild,
                        localeManager.getMessage(ReminderMessages.REMINDERS_EMBED_TITLE),
                        reminder
                    )
                        .setTimestamp(Instant.now())
                        .build()
                )
                .queue()
        }
    }

    private suspend fun dmReminder(guild: Guild, user: Long, reminder: String) {
        val localeManager = LocaleManager[guild]

        GeneralUtils.dmUser(
            user,
            RobertifyEmbedUtils.embedMessageWithTitle(
                guild,
                localeManager.getMessage(ReminderMessages.REMINDERS_EMBED_TITLE),
                reminder
            )
                .setFooter(
                    localeManager.getMessage(
                        ReminderMessages.REMINDER_FROM,
                        Pair("{server}", guild.name)
                    )
                )
                .setTimestamp(Instant.now())
                .build()
        )
    }
}